torch/_utils.py

```diff
@@ -393,3 +393,65 @@ def reraise(self):
             # (https://bugs.python.org/issue2651), so we work around it.
             msg = KeyErrorMessage(msg)
         raise self.exc_type(msg)
+
+
+def _get_available_device_type():
+    if torch.cuda.is_available():
+        return "cuda"
+    # add more available device types here
+    return None
+
+
+def _get_device_attr(get_member):
+    device_type = _get_available_device_type()
+    if device_type.lower() == "cuda":
+        return get_member(torch.cuda)
+    # add more available device types here
+    return None
+
+
+def _get_current_device_index():
+    # current device index
+    return _get_device_attr(lambda m: m.current_device())
+
+
+def _get_all_device_indices():
+    # all device index
+    return _get_device_attr(lambda m: list(range(m.device_count())))
+
+
+def _get_devices_properties(device_ids):
+    # all device properties
+    return [_get_device_attr(lambda m: m.get_device_properties(i)) for i in device_ids]
```

Comment:
(not requesting changes to this PR) The above three functions means that `current_device`, `device_count`, and `get_device_properties` must be available on the module. Do we have any way to enforce that? Say I want to introduce a new device type, what are the expectations? (e.g., which functions I should implement, what invariant I should preserve, etc.)